src/Common/src/System/Net/Logging/GlobalLog.cs

```diff
@@ -0,0 +1,287 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Globalization;
+using System.Threading;
+
+namespace System.Net
+{
+    internal static class GlobalLog
+    {
+        [ThreadStatic]
+        private static Stack<ThreadKinds> t_ThreadKindStack;
+
+        private static Stack<ThreadKinds> ThreadKindStack
+        {
+            get
+            {
+                if (t_ThreadKindStack == null)
+                {
+                    t_ThreadKindStack = new Stack<ThreadKinds>();
+                }
+                return t_ThreadKindStack;
+            }
+        }
+
+        internal static ThreadKinds CurrentThreadKind
+        {
+            get
+            {
+                return ThreadKindStack.Count > 0 ? ThreadKindStack.Peek() : ThreadKinds.Other;
+            }
+        }
+
+        internal static IDisposable SetThreadKind(ThreadKinds kind)
+        {
+            if ((kind & ThreadKinds.SourceMask) != ThreadKinds.Unknown)
+            {
+                throw new InternalException();
+            }
+
+            // Ignore during shutdown.
+            if (Environment.HasShutdownStarted)
+            {
+                return null;
+            }
+
+            ThreadKinds threadKind = CurrentThreadKind;
+            ThreadKinds source = threadKind & ThreadKinds.SourceMask;
+
+            // Special warnings when doing dangerous things on a thread.
+            if ((threadKind & ThreadKinds.User) != 0 && (kind & ThreadKinds.System) != 0)
+            {
+                EventSourceLogging.Log.WarningMessage("Thread changed from User to System; user's thread shouldn't be hijacked.");
+            }
+
+            if ((threadKind & ThreadKinds.Async) != 0 && (kind & ThreadKinds.Sync) != 0)
+            {
+                EventSourceLogging.Log.WarningMessage("Thread changed from Async to Sync, may block an Async thread.");
+            }
+            else if ((threadKind & (ThreadKinds.Other | ThreadKinds.CompletionPort)) == 0 && (kind & ThreadKinds.Sync) != 0)
+            {
+                EventSourceLogging.Log.WarningMessage("Thread from a limited resource changed to Sync, may deadlock or bottleneck.");
+            }
+
+            ThreadKindStack.Push(
+                (((kind & ThreadKinds.OwnerMask) == 0 ? threadKind : kind) & ThreadKinds.OwnerMask) |
+                (((kind & ThreadKinds.SyncMask) == 0 ? threadKind : kind) & ThreadKinds.SyncMask) |
+                (kind & ~(ThreadKinds.OwnerMask | ThreadKinds.SyncMask)) |
+                source);
+
+            if (CurrentThreadKind != threadKind)
+            {
+                Print("Thread becomes:(" + CurrentThreadKind.ToString() + ")");
+            }
+
+            return new ThreadKindFrame();
+        }
+
+        private class ThreadKindFrame : IDisposable
```

Comment:
I'm not sure how this is actually used, as it appears to be here in support of other libraries rather than System.Net.Primitives, but should this be made internal, made a struct, and then SetThreadKind's return type could be changed to ThreadKindFrame rather than IDisposable? (The value SetThreadKind returns would also need to be changed slightly, since ThreadKindFrame couldn't define such a parameterless ctor.)  That would avoid the frame object allocation on each usage of SetThreadKind, though I don't know if/how important that is, since I'm not sure what the use looks like.